### PR TITLE
#170947664 - Make One-way Trip Request

### PR DESCRIPTION
**Pull request by Moise1**
#### What does this PR do?

 It helps a user create a one-way trip request. 

#### Description of Task to be completed?

- Create a one-way trip request
- Have a specific manager to accept/reject the request
- A destination cannot be the same as the departure place
- Cannot make a similar trip twice (same departure date & destination)

#### How should this be manually tested?
- run `git clone https://github.com/andela/knights-bn-backend.git` to clone the repo.
- run `cd knights-bn-backend` to go to the project directory.
- run `git checkout ft-one-way-trip-170947664` to switch to this branch.
- run `npm install` to install dependencies.
- create a Postgres databases locally with PG Admin or Postico
- run `createTables` to create your database. 

- Copy & paste the following in your  **_.env_**  file and fill it accordingly.:  
```
DB_USER='name of the user of the database'
DB_PASS='password of your Postgres'
DB_NAME='name of the database || Barefoot'
DB_TEST_NAME='name of the database for test || Barefoot_test'
NODE_ENV='development'
SECRETKEY='PrivateKey'
BN_API_KEY='use the SendGrid given ken'
MANAGER_PSWD=$2b$10$SO7hwwylEnHKZIv68w/ePOK4yybPnhXa5.tQSUBe.PSb6FBKyuVO.`
HOST_NAME='knights-bn-backend-staging.herokuapp.com'`
LOCAL_HOST='http://localhost:4000' 
```
-  run `npm dev-start` to start the server.
- open postman and do sign up to acquire the token `POST` `http://localhost:4000/api/v1/auth/signup`
  and Fill in the required body request as below.

 ```
{
"firstName": "newiyonsenga",
"lastName": "Eric",
"gender": "male",
"passportNumber": "32825662",
"email": "ishitwill@gmail.com",
"password": "1#Niyonkuru"
}
```

- Do `POST` on `http://localhost:4000/api/v1/trips/oneWayTrip`

```     
{
    "origin": "Kiyovu",
    "destination" :"Gatsata",
   "departureDate" :"2020-03-09",
   "accommodation": "YFB campus",
   "reason": "Attend a meeting"
}
```

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#170947664](https://www.pivotaltracker.com/story/show/170947664)

#### Screenshots
- Sample Response Body 
Screen Shot 2020-03-03 at 22 51 41

### CODE REVIEW

**Comment by Niyonsengaeric** [comment]
@Moise1  but I don't see the reason why you enter your passportNumber on creating the trip while you have entered it on sign up

**Comment by Niyonsengaeric** [comment]
Screen Shot 2020-03-03 at 16 26 38

You can also work again on validations, it saves empty spaces

**Comment by Moise1** [comment]
@Niyonsengaeric I have applied your feedback.

**Comment by Niyonsengaeric** [comment]
> @Niyonsengaeric I have applied your feedback.

Cool let me try it again

**Review by Niyonsengaeric — APPROVED** [review]

**Comment by T2Wil** [comment]
Hey @Moise1 , I think it would be better If we remove some of the non-relevant info for a one_way trip request. I mean returnDate and cities.

image

**Comment by Moise1** [comment]
@T2Wil , please check I've applied your feedback.

**Review by T2Wil — APPROVED** [review]

File: src/controllers/request.js [hunk 1]
```diff
@@ -9,10 +9,56 @@ import { echoNotification } from '../helpers/notificationSender';
...
 export default class usersController {
+  
+  static async createOneWayTrip(req, res) {
+    try {
+      const { id } = req.user;
+      const {
+        origin, destination, departureDate, reason, accommodation
+      } = req.body;
+
+      const managerId = await models.User.findOne({
```
Comment (kagabof) [inline src/controllers/request.js]: You should not use await a then at the same time.

File: src/controllers/request.js [hunk 2]
```diff
@@ -9,10 +9,56 @@ import { echoNotification } from '../helpers/notificationSender';
...
+            { departureDate: Date.parse(departureDate)},
+            { destination }
+          ]
+        }
+      });
+
+      if (existingRequest.length != 0) {
+        return res.status(409).json({ error: 'Sorry! This request already exists. Please double-check your departure date and destination.' })
+      } else {
+        const newTripRequest = await models.Request.create({
```
Comment (kagabof) [inline src/controllers/request.js:36]: You should not use await a then at the same time.

File: src/middlewares/validateOneWay.js [hunk 3]
```diff
@@ -4,29 +4,34 @@ export default async (req, res, next) => {
...
       else return false;
     }
     const destinationCondition = isOriginDiffersFromDestination(origin, destination);
+    const today = new Date;
     const { error } = schema.validate({ origin, destination, departureDate, reason, accommodation });
     if (error) {
       return res.status(422).json({
         status: res.statusCode,
-        error: error.message,
+        Error: error.message,
```
Comment (kagabof) [inline src/middlewares/validateOneWay.js]: why did you change error to Error?

File: src/routes/trips.js [hunk 4]
```diff
@@ -204,30 +282,24 @@
...
-} = requestControllers;
-
 const {
   filterTrips
 } = requestsController;
 
-router.get('/trips/myRequest', authCheck.auth, request.findAllMyRequest);
+const {
+  createTwoWayTrip, pendingApproval, createOneWayTrip, rejectRequest, createMultiCityRequest
+} = requestControllers;
```
Comment (kagabof) [inline src/routes/trips.js]: @Moise1, I don't understand why initializing new constant variables while you can destructure by the time you are importing.

File: src/controllers/request.js (hunk 1 above)
Comment (kagabof) [inline src/controllers/request.js]: You should also add this in helpers for avoiding code redundancy and for the future use.

**Review by kagabof — CHANGES_REQUESTED** [review]

**Review by kagabof — CHANGES_REQUESTED** [review]
Please check your commit message and follow the  https://github.com/andela/engineering-playbook/tree/master/5.%20Developing/Conventions#commit-message

**Comment by Moise1** [comment]
@kagabof  please take a look at the commit message. I've applied the feedback.

**Review by kagabof — APPROVED** [review]
Good job.

**Pull request merged**